### PR TITLE
Add `--version` flag

### LINKED ISSUE #11241: implement servo --version , and include the compilation date
In the browserhtml team, we use servo nightly builds, and it's always difficult to know which version of servo one is currently using.

Could we have a `--version` option, that would return something like:

```
$ servo --version
Servo 0.0.1 - compiled 2016-05-27
```

### PR DESCRIPTION
**Pull request by campaul**
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy --faster` does not report any errors
- [X] These changes fix #11241 (github issue number if applicable).

Either:
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

Not 100% sure of a good way to test this, so I'm submitting as is for feedback. Manually testing it appears to work fine.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11262)

### CODE REVIEW

**Comment by Manishearth** [comment]
Could we also include the commit hash (`git rev-parse HEAD`)?

**Comment by aneeshusa** [comment]
Please do not include the BUILD_DATE - this [kills reproducibility](https://reproducible-builds.org/docs/timestamps/).

We could use the [SOURCE_DATE_EPOCH](https://reproducible-builds.org/specs/source-date-epoch/) environment variable instead, but I'd rather just not have the timestamp at all.

**Comment by Manishearth** [comment]
Oh, good point. I'd prefer to just have a git hash in that case.

**Comment by paulrouget** [comment]
git hash works for me.

**Comment by campaul** [comment]
I agree that including the build date is problematic. The build date has no real correlation with version (I could check out a really old version and build today for example). It might be better to include the git hash instead of the date, rather than in addition to.

**Comment by metajack** [comment]
If you are going to include a hash, it might be nice to include whether the tree was dirty when built. This has been very useful in the Opus and Daala projects to let us know when bug reports come in with modifications from the source tree. See https://github.com/xiph/daala/search?utf8=%E2%9C%93&q=dirty for example.

**Comment by campaul** [comment]
Here is all the different data that can be easily accessed using this PR as an example:

- cargo package version: `0.0.1`
- git description: `homu-tmp-1916-g3b1b90a` or `homu-tmp-1916-g3b1b90a-dirty`
- git sha: `3b1b90ab90a315d959f41b2e821158a547818751`
- git sha (short): `3b1b90a`

Right now the only tag that exists is `homu-tmp`, but if there were version tags then git description would look like `v0.0.1-1916-g3b1b90a`.

So given all of that as possible input, is there a consensus on what the output of `--version` should look like?

**Comment by Manishearth** [comment]
We don't use tags yet, perhaps `v0.0.1-3b1b90a` using the cargo version and the short sha?

**Comment by campaul** [comment]
Should that also include an optional `-dirty` suffix, so you get `v0.0.1-3b1b90a-dirty` when the working directory isn't clean?

**Comment by aneeshusa** [comment]
+1 on optional dirty suffix.

Also, please make the git-related info optional; this is much nicer for packaging because you can build from a tarball instead of needing the full git history (or a shallow clone). (I imagine that in the future, `./mach package` may use a different method to inject the current version if necessary.)

**Comment by campaul** [comment]
The updated CR has the following behavior:

- `Servo 0.0.1` if you're not in a git repo
- `Servo 0.0.1-<sha>` if you're in a git repo
- `Servo 0.0.1-<sha>-dirty` if you're in a git repo with uncommitted changes

I moved the logic for printing the version into `components/servo/main.rs` so that it prints the version of servo and not the version of the util package. It's more complicated, but it's actually correct now.

The only behavior that I'm unsure of is what to do when changes occur outside of source files. For example, if I build with uncommitted changes, then commit those changes, then try to build again nothing will happen because no rust file has changed, and the servo binary will still say `-dirty` even though the working directory is clean.

**Comment by jdm** [comment]
ping @Manishearth!

File: components/servo/main.rs [hunk 1]
```diff
@@ -35,12 +35,18 @@ use servo::Browser;
...
 use servo::util::opts::{self, ArgumentParsingResult};
 use servo::util::panicking::initiate_panic_hook;
+use std::process;
 use std::rc::Rc;
 
 fn main() {
     // Parse the command line options and store them globally
     let opts_result = opts::from_cmdline_args(&*args());
 
+    if opts::get().is_printing_version {
```
Comment (Ms2ger) [inline components/servo/main.rs]: Move this below the `if let Some(token) = content_process_token {` block below.

**Comment by paulrouget** [comment]
Any update?

**Comment by campaul** [comment]
Sorry, I've been really busy. I'll make the requested change tonight and update the PR.

**Pull request merged**